### PR TITLE
xl/ReadFile: code refactor. Fix for read-failure if one of the parts …

### PR DESCRIPTION
**Pull request by krishnasrinivas**
fixes crash during ReadFile if one of the parts is missing.
Also refactor code to use the apis getMetadata() setMetadata()

### CODE REVIEW

**Comment by balamurugana** [comment]
:-1:
1. Why do you want to refactor to fix the crash?  Can you also point here how the crash looks like?
2. setMetadata() and getMetadata() should be a separate patch.

File: xl-v1-readfile.go [hunk 1]
```diff
@@ -141,43 +141,23 @@ func (xl XL) ReadFile(volume, path string, offset int64) (io.ReadCloser, error)
...
 
 	readers := make([]io.ReadCloser, len(quorumDisks))
-	readFileError := 0
-	for _, quorumDisk := range quorumDisks {
-		erasurePart := slashpath.Join(path, fmt.Sprintf("part.%d", quorumDisk.index))
-		var erasuredPartReader io.ReadCloser
-		if erasuredPartReader, err = quorumDisk.disk.ReadFile(volume, erasurePart, offset); err != nil {
-			// We can safely allow ReadFile errors up to len(quorumDisks) - xl.readQuorum
-			// otherwise return failure
-			if readFileError < len(quorumDisks)-xl.readQuorum {
```
Comment (balamurugana) [inline xl-v1-readfile.go:164]: Isn't this helping to avoid missing part file error?

**Comment by krishnasrinivas** [comment]
> Why do you want to refactor to fix the crash? Can you also point here how the crash looks like?

I wanted refactor so that setMetadata() and getMetadata() can be used by both xl.ReadFile and xl.selfHeal code so that there is no repetition of code. But only during testing I realized that xl.ReadFile crashes if a part is missing.

Here's the crash:

```
2016/04/20 23:39:30 http: panic serving 127.0.0.1:47916: runtime error: index out of range
goroutine 29 [running]:
net/http.(*conn).serve.func1(0xc820226000)
        /home/krishna/go/src/net/http/server.go:1389 +0xc1
panic(0x9e0840, 0xc82000e0f0)
        /home/krishna/go/src/runtime/panic.go:426 +0x4e9
main.XL.ReadFile(0x7f58e72d7200, 0xc820014820, 0x8, 0x8, 0xc820260200, 0x10, 0x10, 0xc8202291d0, 0xc82000e980, 0x9, ...)
        /home/krishna/gocode/src/github.com/minio/minio/xl-v1-readfile.go:180 +0xd3f
main.(*XL).ReadFile(0xc8202341e0, 0xc82038c925, 0x5, 0xc82038c92b, 0x6, 0x0, 0x0, 0x0, 0x0, 0x0)
        <autogenerated>:173 +0x113
main.objectAPI.GetObject(0x7f58e72d72b0, 0xc8202341e0, 0xc82038c925, 0x5, 0xc82038c92b, 0x6, 0x0, 0x0, 0x0, 0x6)
        /home/krishna/gocode/src/github.com/minio/minio/object-api.go:128 +0x270
main.objectAPIHandlers.GetObjectHandler(0x7f58e72d72b0, 0xc8202341e0, 0x7f58e73270a8, 0xc8201a29c0, 0xc820478c40)
        /home/krishna/gocode/src/github.com/minio/minio/object-handlers.go:138 +0x87d
main.(objectAPIHandlers).GetObjectHandler-fm(0x7f58e73270a8, 0xc8201a29c0, 0xc820478c40)
        /home/krishna/gocode/src/github.com/minio/minio/api-router.go:49 +0x4a
net/http.HandlerFunc.ServeHTTP(0xc820295d20, 0x7f58e73270a8, 0xc8201a29c0, 0xc820478c40)
        /home/krishna/go/src/net/http/server.go:1618 +0x3a
github.com/minio/minio/vendor/github.com/gorilla/mux.(*Router).ServeHTTP(0xc820014d70, 0x7f58e73270a8, 0xc8201a29c0, 0xc820478c40)
        /home/krishna/gocode/src/github.com/minio/minio/vendor/github.com/gorilla/mux/mux.go:100 +0x29e
main.redirectHandler.ServeHTTP(0x7f58e72d7690, 0xc820014d70, 0xa94770, 0x6, 0x7f58e73270a8, 0xc8201a29c0, 0xc820478c40)
        /home/krishna/gocode/src/github.com/minio/minio/generic-handlers.go:71 +0x278
main.(*redirectHandler).ServeHTTP(0xc8202bf1c0, 0x7f58e73270a8, 0xc8201a29c0, 0xc820478c40)
        <autogenerated>:84 +0xb4
main.minioPrivateBucketHandler.ServeHTTP(0x7f58e72d76b8, 0xc8202bf1c0, 0xa94770, 0x6, 0x7f58e73270a8, 0xc8201a29c0, 0xc820478c40)
        /home/krishna/gocode/src/github.com/minio/minio/generic-handlers.go:119 +0x1a5
main.(*minioPrivateBucketHandler).ServeHTTP(0xc8202bf1e0, 0x7f58e73270a8, 0xc8201a29c0, 0xc820478c40)
        <autogenerated>:86 +0xb4
main.cacheControlHandler.ServeHTTP(0x7f58e72d76e0, 0xc8202bf1e0, 0x7f58e73270a8, 0xc8201a29c0, 0xc820478c40)
        /home/krishna/gocode/src/github.com/minio/minio/generic-handlers.go:100 +0x26a
main.(*cacheControlHandler).ServeHTTP(0xc8202bbb10, 0x7f58e73270a8, 0xc8201a29c0, 0xc820478c40)
        <autogenerated>:85 +0xb6
```

There were also other issues like:
- `fileSize, err := xl.getFileSize(volume, path, quorumDisks[0].disk)`
  if this fails we don't take care of trying it on other nodes, this not a big thing but the current patch takes care of it. Also there is no need of doing xl.getFileSize as we would have already fetched metadata info during `xl.getReadFileQuorumDisks`.

*

```
                        // Loop through all readers and read.                                                                                                                                                                        
                        for index, reader := range readers {
                                if reader == nil {
                                        // One of files missing, save it for reconstruction.                                                                                                                                         
                                        enBlocks[index] = nil
                                        continue
                                }
                                // Initialize shard slice and fill the data from each parts.                                                                                                                                         
                                enBlocks[index] = make([]byte, curEncBlockSize)
                                _, err = io.ReadFull(reader, enBlocks[index])
                                if err != nil && err != io.ErrUnexpectedEOF {
                                        enBlocks[index] = nil
                                }
                        }
```

ReedSolomon.Verify() expects that enBlocks[index] is not nil, else it returns `ErrShardSize`

> setMetadata() and getMetadata() should be a separate patch.

My main intention was to refactor to have setMetadata and getMetadata so that it can be used by both ReadFile and selfHeal, crash bug fix was detected during testing after refactor and hence fixed that too.

File: xl-v1-readfile.go (hunk 1 above)
Comment (krishnasrinivas) [inline xl-v1-readfile.go:164]: this code also has the bug when len(quorumDisks) < xl.readQuorum

**Comment by harshavardhana** [comment]
This patch is squashed and merged.

File: xl-v1-readfile.go (hunk 1 above)
Comment (balamurugana) [inline xl-v1-readfile.go:164]: this case is handled above like

``` go
    // Check read quorum.
    quorumDisks, _ := xl.getQuorumDisks(volume, path)
    if len(quorumDisks) < xl.readQuorum {
        return nil, errReadQuorum
    }
```

**Pull request closed without merging**